src/Aspire.Hosting/ResourceBuilderExtensions.cs

```diff
@@ -549,4 +551,82 @@ public static IResourceBuilder<T> ExcludeFromManifest<T>(this IResourceBuilder<T
+            var resourceLogger = rls.GetLogger(builder.Resource);
+            resourceLogger.LogInformation("Waiting for resource '{Name}' to enter the '{State}' state.", dependency.Resource.Name, KnownResourceStates.Running);
+
+            var rns = e.Services.GetRequiredService<ResourceNotificationService>();
+            await rns.PublishUpdateAsync(builder.Resource, s => s with { State = KnownResourceStates.Waiting }).ConfigureAwait(false);
+            await rns.WaitForResourceAsync(dependency.Resource.Name, cancellationToken: ct).ConfigureAwait(false);
+        });
+
+        return builder;
+    }
+
+    /// <summary>
+    /// Waits for the dependency resource to enter the Exited or Finished state before starting the resource.
+    /// </summary>
+    /// <typeparam name="T">The type of the resource.</typeparam>
+    /// <param name="builder">The resource builder for the resource that will be waiting.</param>
+    /// <param name="dependency">The resource builder for the dependency resource.</param>
+    /// <returns>The resource builder.</returns>
+    /// <remarks>
+    /// <para>This method is useful when a resource should wait until another has completed. A common usage pattern
+    /// would be to include a console application that initializes the database schema or performs other one off
+    /// initialization tasks.</para>
+    /// <para>Note that this method has no impact at deployment time and only works for local development.</para>
+    /// </remarks>
+    /// <example>
+    /// Wait for database initialization app to complete running.
+    /// <code lang="C#">
+    /// var builder = DistributedApplication.CreateBuilder(args);
+    /// var pgsql = builder.AddPostgres("postgres");
+    /// var dbprep = builder.AddProject&lt;Projects.DbPrepApp&gt;("dbprep")
+    ///                     .WithReference(pgsql);
+    /// builder.AddProject&lt;Projects.DatabasePrepTool&gt;("dbprep")
+    ///        .WithReference(pgsql)
+    ///        .WaitForCompletion(dbprep);
+    /// </code>
+    /// </example>
+    public static IResourceBuilder<T> WaitForCompletion<T>(this IResourceBuilder<T> builder, IResourceBuilder<IResource> dependency) where T : IResource
+    {
+        builder.ApplicationBuilder.Eventing.Subscribe<BeforeResourceStartedEvent>(builder.Resource, async (e, ct) =>
+        {
+            var rls = e.Services.GetRequiredService<ResourceLoggerService>();
+            var resourceLogger = rls.GetLogger(builder.Resource);
+            resourceLogger.LogInformation($"Waiting for resource '{dependency.Resource.Name}' to enter the 'Finished' state.");
+
+            var rns = e.Services.GetRequiredService<ResourceNotificationService>();
+            await rns.PublishUpdateAsync(builder.Resource, s => s with { State = KnownResourceStates.Waiting }).ConfigureAwait(false);
+            await rns.WaitForResourceAsync(dependency.Resource.Name, targetState: "Finished", cancellationToken: ct).ConfigureAwait(false);
```

Comment:
Not sure this is good enough.

https://github.com/davidfowl/WaitForDependenciesAspire/blob/389f02daf2a3903924875c601dc300371aa99e27/WaitForDependencies.Aspire.Hosting/WaitForDependenciesExtensions.cs#L140

Reply:
I think we should wait for `Exited` - but I am not sure about the best way of handling `FailedToStart` especially in the context of container restart scenarios.

Reply:
It should fail the wait.

Reply:
OK this is fixed up. In the end I added an overload to the ResourceNotificationService which allows you to get the snapshot back and also provide a predicate (just makes the API easier to use rather than having to do your own wait loop).

Once I've got the snapshot I check for the exit code and if it is not the target (default 0) then it blows up which blocks the dependent resource starting. Notice that the `WaitForCompletion` only sets the state to Waiting it doesn't set its subsequent state.

Its up for the code firing off the events to interpret what an exception in any of the event subscribers means. For this event it stops the startup of the resource.